sources/engine/Xenko.Graphics/Sprite3DBatch.cs

```diff
@@ -142,41 +143,58 @@ protected override unsafe void UpdateBufferValuesFromElementInfo(ref ElementInfo
-                var textureCoordY = new Vector2(drawInfo->Source.Top, drawInfo->Source.Bottom);
-
-                // set the two first line of vertices
-                for (int r = 0; r < 2; r++)
+                const int VertexCount = 4;
+                var texCoords = stackalloc Vector2[]
                 {
-                    for (int c = 0; c < 2; c++)
-                    {
-                        vertex->ColorScale = drawInfo->ColorScale;
-                        vertex->ColorAdd = drawInfo->ColorAdd;
-
-                        vertex->Swizzle = (int)drawInfo->Swizzle;
-                        vertex->TextureCoordinate.X = textureCoordX[c];
-                        vertex->TextureCoordinate.Y = textureCoordY[r];
-
-                        vertex->Position.X = currentPosition.X;
-                        vertex->Position.Y = currentPosition.Y;
-                        vertex->Position.Z = currentPosition.Z;
-                        vertex->Position.W = currentPosition.W;
-
-                        vertex++;
-
-                        if (c == 0)
-                            Vector4.Add(ref currentPosition, ref drawInfo->UnitXWorld, out currentPosition);
-                        else
-                            Vector4.Subtract(ref currentPosition, ref drawInfo->UnitXWorld, out currentPosition);
-                    }
-
-                    Vector4.Add(ref currentPosition, ref drawInfo->UnitYWorld, out currentPosition);
+                    drawInfo->Source.TopLeft,
+                    drawInfo->Source.TopRight,
+                    drawInfo->Source.BottomLeft,
+                    drawInfo->Source.BottomRight,
+                };
+
+                ref var startPos = ref drawInfo->LeftTopCornerWorld;
+                var vertexPositions = stackalloc Vector4[]
+                {
+                    startPos,                                                                       // Top Left
+                    Vector4Add(ref startPos, ref drawInfo->UnitXWorld),                             // Top Right
+                    Vector4Add(ref startPos, ref drawInfo->UnitYWorld),                             // Bottom Left (Y axis points up, but Y value will be negative value to orientate correctly)
+                    Vector4Add(ref startPos, ref drawInfo->UnitXWorld, ref drawInfo->UnitYWorld),   // Bottom Right
+                };
+
+                var swizzle = (float)drawInfo->Swizzle;
+                for (int i = 0; i < VertexCount; i++, vertex++)
+                {
+                    vertex->ColorScale = drawInfo->ColorScale;
+                    vertex->ColorAdd = drawInfo->ColorAdd;
+                    vertex->Swizzle = swizzle;
+                    vertex->TextureCoordinate = texCoords[i];
+                    vertex->Position = vertexPositions[i];
                 }
             }
         }
-         
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Not sure it does anything. Inlining rules are pretty strict and most of the time this has no effect. Unless we can prove it does something, we should probably leave it out as it otherwise would be misleading.

According to some sources, it seems to be the case for methods having structs as parameter. But here, since the values are given by reference, I don't know if it still applies. By even if it did, I think the method body is still too big to be considered for inlining.

Reply:
I think you're right. I tried to benchmark one with AggressiveInlining and one without and I think the numbers were generally the same. I've removed it now.